SocialTradingManager and Allocator

bweick opened the pull request.

felix2feng commented on artifacts/index.ts:
```diff
@@ -57,4 +58,5 @@ export {
 	RSITrendingTriggerContract,
 	TimeSeriesFeedContract,
 	TwoAssetLinearizedTimeSeriesFeedContract,
```
Export the `SocialTradingManager` too?

bweick replied:
Weird I thought I updated those. But yes +1

felix2feng commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,296 @@
+     */
+    function createTradingPool(
+        ISocialAllocator _tradingPairAllocator,
+        address _factory,
```
Why choose not to hardcode factory hard-coded in the state of this contract?

bweick replied:
Potential for different RebalancingSet variations

felix2feng commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,296 @@
+    /* ============ State Variables ============ */
+
+    ICore public core;
+    mapping(address => PoolInfo) public pools;
```
Maybe you can do `IRebalancingSetTokenV2` => `PoolInfo` for more explicitness

bweick replied:
Nope tried it. Not allowed in Solidity

felix2feng commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,296 @@
+     * @param _newAllocation      New base asset allocation in a scaled decimal value
+     *                                          (e.g. 100% = 1e18, 1% = 1e16)
+     */
+    function updateAllocation(
```
What's the gas usage of calling this function / how much does it cost to make a trade?

bweick replied:
`1725208` for reference creating a `tradingPool` is `5108987` which is not terribly cheap

felix2feng commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,296 @@
+    {
+        validateAllocationUpdate(_tradingPool, _newAllocation);
+
+        ISetToken nextSet = pools[address(_tradingPool)].allocator.determineNewAllocation(
```
May be cleaner to have a `allocator(_tradingPool)` function.

Allocator would return `pool[address(_tradingPool)].allocator`

bweick replied:
Sure

felix2feng commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,296 @@
+    /* ============ Constants ============ */
+
+    uint256 public constant REBALANCING_SET_NATURAL_UNIT = 10 ** 6;
+    uint256 public constant SCALE_FACTOR = 10 ** 18;
```
Let's consider exposing this in `CommonMath` as a function so it doesn't need to be redeclared every file

bweick replied:
+1

felix2feng commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,296 @@
+            SCALE_FACTOR
+        );
+
+        _tradingPool.startRebalance(address(nextSet));
```
As you know: With the new changes, it has the `liquidatorData` argument.

bweick replied:
+1

felix2feng commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,296 @@
+        // Require that enough time has passed from last rebalance
+        uint256 lastRebalanceTimestamp = _tradingPool.lastRebalanceTimestamp();
+        uint256 rebalanceInterval = _tradingPool.rebalanceInterval();
+        require(
```
Not opposed to these checks being here - though am noting they are on the RBSetV2 already

bweick replied:
Yeah it's something we did on the previous managers to stop excess gas usage. This way the user doesn't accidentally call it and then is responsible for the gas cost of creating a SetToken

felix2feng commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,296 @@
+    event TradingPoolCreated(
+        address indexed _trader,
+        ISocialAllocator indexed _allocator,
+        address _tradingPool,
```
This could just have a type of `IRebalancingSetTokenV2` - since there wouldn't need to be conversions later

bweick replied:
Well this is emitted when we call create on `Core` which returns an address. And since we can't save in the mapping as IRebalancingSetTokenV2 defining the output here as such would actually be converting something that doesn't really need to be converted.

felix2feng commented on contracts/managers/allocators/SocialAllocator.sol:
```diff
@@ -0,0 +1,401 @@
+            nextSetComponents,
+            nextSetUnits,
+            collateralNaturalUnit,
+            "",
```
When debugging / searching on etherscan, it's actually nice for the collateral set to have a name/symbol. 

At the very least, maybe this could be passed in thru the constructor

bweick replied:
Guess that was a part I left unfilled. What names would you propose?

felix2feng commented on contracts/managers/allocators/SocialAllocator.sol:
```diff
@@ -0,0 +1,401 @@
+        view
+        returns (uint256)
+    {
+        return calculateCollateralSetValueInternal(
```
Style: Could have this in one line

felix2feng commented on contracts/managers/allocators/SocialAllocator.sol:
```diff
@@ -0,0 +1,401 @@
+     * @param  _currentCollateralSet        Instance of current set collateralizing RebalancingSetToken
+     * @return uint256                      USD value of passed Set
+     */
+    function calculateCollateralSetValueInternal(
```
In `set-protocol-contracts`, there's a `SetUSDValuation` file. Perhaps we can standardize the code for calculating USD price of a Set, so we don't need to rewrite/reimplement across codebases.

The other use that seems to clean things up is also using `oracleWhitelist` vs. individual oracleAddresses passed in the constructor and held in state

felix2feng commented on contracts/managers/allocators/SocialAllocator.sol:
```diff
@@ -0,0 +1,401 @@
+     */
+    function calculateNextSetParameters(
+        uint256 _targetBaseAssetAllocation,
+        uint256 _allocationPrecision
```
Why do we need to pass `allocationPrecision` again vs hardcoding `SCALE_FACTOR`?

bweick replied:
Was done more to be explicit and allow flexibility. But suppose it could be removed in this instance

felix2feng commented on contracts/managers/allocators/SocialAllocator.sol:
```diff
@@ -0,0 +1,401 @@
+        view
+        returns (uint256[] memory, address[] memory)
+    {
+        // Get prices
```
To save gas, these getPrices functions could be in the if statement

bweick replied:
How does that save gas? They're used in both instances.

felix2feng commented on contracts/managers/allocators/SocialAllocator.sol:
```diff
@@ -0,0 +1,401 @@
+     * @return uint256[]                        Unit array of nextSet collateral
+     * @return address[]                        Component array of nextSet collateral
+     */
+    function calculateSingleAssetNextSetUnits(
```
Re: naming, what is returned is actually the allocation (defined as address + units). Function naming should reflect

bweick replied:
How does that save gas? They're used in both instances.

bweick commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,296 @@
+            _startingBaseAssetAllocation,
+            SCALE_FACTOR
+        );
+        
```
Then it would operate like a normal RebalancingSet but the manager would have to come up with the allocations on their own.

bweick commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,296 @@
+            _newAllocation <= SCALE_FACTOR,
+            "SocialTradingManager.validateAllocationUpdate: New allocation is not valid."
+        );
+
```
Is there a minimum allocation amount we want to enforce? I guess I had 1% in mind which would be 1e16. I'd be fine with that actually.

bweick commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,296 @@
+
+        pools[address(_tradingPool)].trader = _newTrader;        
+    }
+
```
Yeah problem with that is that they would be able to then switch assets too. Unless we're ok with that?

bweick commented on contracts/managers/allocators/SocialAllocator.sol:
```diff
@@ -0,0 +1,401 @@
+    address[] public multiAssetNextSetComponents;
+    address[] public baseAssetNextSetComponents;
+    address[] public quoteAssetNextSetComponents;
+    uint256 public baseAssetDecimalDifference;
```
Feel like in our system we've just been using decimal as the full unit nomenclature too. But yes I suppose there is that difference.

bweick commented:
Internal Audit Fixes CHANGELOG:
- _allocationPrecision removed from determineNewAllocation interface
- Remove SCALE_FACTOR constant and assume uints are using 18 decimals
- Added liquidatorData to startRebalance call and updateAllocation interface
- Added separate function to validate allocation, no allocation greater than `10**18` or less than `10**16` (except 0) allowed
- Added helper functions for interaction with pools struct
- SocialAllocator constructor now accepts `oracleWhiteList` , `collateralName`, and `collateralSymbol` params
- SetUSDValuation.calculateSetTokenDollarValue used to calculate collateral USD Value

felix2feng commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,323 @@
+                _allocation <= CommonMath.scaleFactor(),
+                "Passed allocation must not exceed 100%."
+            );
+
```
Would we also want it to be a multiple of 10**16?

felix2feng commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,323 @@
+
+    modifier onlyTrader(IRebalancingSetTokenV2 _tradingPool) {
+        require(
+            msg.sender == pools[address(_tradingPool)].trader,
```
Update `trader` with new construct

felix2feng commented on contracts/managers/allocators/SocialAllocator.sol:
```diff
@@ -0,0 +1,364 @@
+        // Calculate multiplier for quote and base asset. Multiplier is just the amount of highest
+        // allocation divided by lowest allocation. Asset that has lowest allocation will have
+        // multiplier set to 1.
+        uint256 baseAssetMultiplier = Math.max(_targetBaseAssetAllocation.scale().div(quoteAssetAllocation), ONE.scale());
```
Why `ONE.scale()` vs `CommonMath.scaleFactor`?

felix2feng commented on contracts/managers/SocialTradingManager.sol:
```diff
@@ -0,0 +1,323 @@
+    struct PoolInfo {
+        address trader;
+        ISocialAllocator allocator;
+        uint256 currentAllocation;
```
Since we don't have a type for scaledValues, I've been suffixing all `scaled` values with `Scaled`. So in this example, it'd be `currentAllocationScaled`
